cpp/unreal_plugins/SimulationController/Source/SimulationController/WheeledVehicleAgent.cpp

```diff
@@ -0,0 +1,310 @@
+}
+
+std::map<std::string, ArrayDesc> WheeledVehicleAgent::getObservationSpace() const
+{
+    std::map<std::string, ArrayDesc> observation_space;
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.WHEELED_VEHICLE_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "state_data")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<float>::lowest();
+        array_desc.high_ = std::numeric_limits<float>::max();
+        array_desc.datatype_ = DataType::Float64;
+        array_desc.shape_ = { 6 };
+        observation_space["state_data"] = std::move(array_desc); // position (X, Y, Z) and orientation (Roll, Pitch, Yaw) of the agent relative to the world frame.
+    }
+
+    if (Std::contains(observation_components, "encoder")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<float>::lowest();
+        array_desc.high_ = std::numeric_limits<float>::max();
+        array_desc.datatype_ = DataType::Float64;
+        array_desc.shape_ = { 4 };
+        observation_space["encoder"] = std::move(array_desc); // FL, FR, RL, RR
+    }
+
+    if (Std::contains(observation_components, "imu")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<float>::lowest();
+        array_desc.high_ = std::numeric_limits<float>::max();
+        array_desc.datatype_ = DataType::Float64;
+        array_desc.shape_ = { 6 };
+        observation_space["imu"] = std::move(array_desc); // a_x, a_y, a_z, g_x, g_y, g_z
+    }
+
+    if (Std::contains(observation_components, "sonar")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<float>::lowest();
+        array_desc.high_ = std::numeric_limits<float>::max();
+        array_desc.datatype_ = DataType::Float64;
+        array_desc.shape_ = { 1 };
+        observation_space["sonar"] = std::move(array_desc); // Front obstacle distance in [m]
+    }
+
+    std::map<std::string, ArrayDesc> camera_sensor_observation_space = camera_sensor_->getObservationSpace(observation_components);
+    for (auto& camera_sensor_observation_space_component : camera_sensor_observation_space) {
+        observation_space[camera_sensor_observation_space_component.first] = std::move(camera_sensor_observation_space_component.second);
+    }
+
+    return observation_space;
+}
+
+std::map<std::string, ArrayDesc> WheeledVehicleAgent::getStepInfoSpace() const
+{
+    return {};
+}
+
+void WheeledVehicleAgent::applyAction(const std::map<std::string, std::vector<uint8_t>>& action)
+{
+    auto action_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.WHEELED_VEHICLE_AGENT.ACTION_COMPONENTS");
+
+    if (Std::contains(action_components, "apply_wheel_torques")) {
+        std::vector<double> component_data = Std::reinterpretAs<double>(action.at("apply_wheel_torques"));
+        wheeled_vehicle_pawn_->setDriveTorques(Eigen::Vector4d(
```

Comment:
I don't think it's worth it to use Eigen here. Why not just pass in the std::vector<double> directly.